pep-0xxx.rst

```diff
@@ -0,0 +1,1317 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+NSS is Mozilla's crypto and TLS library. It's used in Firefox, Thunderbird,
+and as alternative to OpenSSL in multiple libraries, e.g. curl.
+
+By default, NSS comes with secure configuration of allowed ciphers. On some
+platforms such as Fedora, the list of enabled ciphers is globally configured
+in a system policy. Generally, applications should not modify cipher suites
+unless they have specific reasons to do so.
+
+NSS has both process global and per-connection settings for cipher suites. It
+does not have a concept of SSLContext like OpenSSL. A SSLContext-like behavior
+can be easily emulated. Specifically, ciphers can be enabled or disabled
+globally with ```SSL_CipherPrefSetDefault(PRInt32 cipher, PRBool enabled)```,
+and ```SSL_CipherPrefSet(PRFileDesc *fd, PRInt32 cipher, PRBool enabled)```
+for a connection. The cipher ```PRInt32``` number is a signed 32bit integer
+that directly corresponds to an registered IANA id, e.g. ```0x1301```
+is ```TLS_AES_128_GCM_SHA256```. Contrary to OpenSSL, the preference order
+of ciphers is fixed and cannot be modified at runtime.
+
+Like SecureTransport, NSS has no API for aggregated entries. Some consumers
+of NSS have implemented custom mappings from OpenSSL cipher names and rules
+to NSS ciphers, e.g. ```mod_nss```.
+
+
+Proposed Interface
+^^^^^^^^^^^^^^^^^^
+
+The proposed interface for the new module is influenced by the combined set of
+limitations of the above implementations. Specifically, as every implementation
+*except* OpenSSL requires that each individual cipher be provided, there is no
+option but to provide that lowest-common denominator approach.
+
+The simplest approach is to provide an enumerated type that includes all of the
+cipher suites defined for TLS. The values of the enum members will be their
+two-octet cipher identifier as used in the TLS handshake, stored as a tuple of
+integers. The names of the enum members will be their IANA-registered cipher
+suite names.
+
+Rather than populate this enum by hand, it is likely that we'll define a
+script that can build it from Christian Heimes' `tlsdb JSON file`_ (warning:
+large file). This also opens up the possibility of extending the API with
+additional querying function, such as determining which TLS versions support
+which ciphers, if that functionality is found to be useful or necessary.
+
+If users find this approach to be onerous, a future extension to this API can
+provide helpers that can reintroduce OpenSSL's aggregation functionality.
+
+Because this enum would be enormous, the entire enum is not provided here.
+Instead, a small sample of entries is provided to give a flavor of how it will
+appear.
+
+::
+
+    class CipherSuite(Enum):
+        ...
+        TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA = (0xC0, 0x12)
+        ...
+        TLS_ECDHE_ECDSA_WITH_AES_128_CCM = (0xC0, 0xAC)
+        ...
+        TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 = (0xC0, 0x2B)
+        ...
```

Comment:
I have compiled a list of cipher suites. IANA currently defines 329 ciphers suites. Without deprecated, unsupported and irrelevant cipher suites, I'm down to 115 suites. The list of suites omits

* all export ciphers
* key agreements NULL, SRP_SHA, KRB5, PSK, PSK DHE
* authentication anon, NULL, PSK, DSS
* encryption NULL, IDEA, ARIA

https://gist.github.com/tiran/283658a2084b007cd918067c0cbc77d4